distsqlplan: fake span resolver

RaduBerinde opened the pull request:
Implementation and test of the fake span resolver, which does a key scan in the
queried span and splits it arbitrarily between all the nodes. On average, each
nodes should get around 5 ranges (if there are enough keys).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13186)
<!-- Reviewable:end -->

RaduBerinde committed:
distsqlplan: fake span resolver

Implementation and test of the fake span resolver, which does a key scan in the
queried span and splits it arbitrarily between all the nodes. On average, each
nodes should get around 5 ranges (if there are enough keys).

RaduBerinde committed:
distsqlplan: span resolver interfaces

Making `SpanResolver` and `SpanResolverIterator` into interfaces that will also
be implemented by the fake resolver.

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> all right!

---

Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlplan/fake_span_resolver.go, line 37 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13186#-Kbuu1XqtRhYNIgbffVT:-Kbuu1XrrdN0mhnMKC4r:b6q0huq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b177a8f780ed7791e3de77c0815dfe5a42e5db28/pkg/sql/distsqlplan/fake_span_resolver.go#L37)):*
> ```Go
> var _ SpanResolver = &fakeSpanResolver{}
> 
> // NewFakeSpanResolver creates a SpanResolver which splits spans and distributes
> ```

i'd suggest moving this comment to the definition of the `fakeSpanResolver` struct, and also expanding it a bit to detail what it does - what `avgRangesPerNode` means.
And also state that each `Seek` randomizes everything.

---

*[pkg/sql/distsqlplan/fake_span_resolver.go, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13186#-Kbuucjz7LBm2gaBEVmm:-Kbuucjz7LBm2gaBEVmn:b-pfapow) ([raw file](https://github.com/cockroachdb/cockroach/blob/b177a8f780ed7791e3de77c0815dfe5a42e5db28/pkg/sql/distsqlplan/fake_span_resolver.go#L80)):*
> ```Go
> 	var splitKeys []roachpb.Key
> 
> 	if err := fit.fsr.db.Txn(ctx, func(txn *client.Txn) error {
> ```

why do you need a transaction here? you can do non-transactional scans...

---

*[pkg/sql/distsqlplan/fake_span_resolver.go, line 112 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13186#-KbuvcnqHfzPC68hAgXs:-KbuvcnqHfzPC68hAgXt:b-p17mnc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b177a8f780ed7791e3de77c0815dfe5a42e5db28/pkg/sql/distsqlplan/fake_span_resolver.go#L112)):*
> ```Go
> 
> 	// Generate fake splits.
> 	maxSplits := 2 * len(fit.fsr.nodes) * avgRangesPerNode
> ```

why `2 * ...` ?

---

*[pkg/sql/distsqlplan/fake_span_resolver.go, line 118 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13186#-Kbuv8Umd2KkdBmudzWM:-Kbuv8Umd2KkdBmudzWN:b32vin) ([raw file](https://github.com/cockroachdb/cockroach/blob/b177a8f780ed7791e3de77c0815dfe5a42e5db28/pkg/sql/distsqlplan/fake_span_resolver.go#L118)):*
> ```Go
> 	numSplits := rand.Intn(maxSplits)
> 
> 	// Use Robert Floyd's algorithm to generate numSplits distinct integers
> ```

:D

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13186#-:-KbuwMybSBReC5qRrbJ5:b-15zast)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!

---

Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlplan/fake_span_resolver.go, line 37 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13186#-Kbuu1XqtRhYNIgbffVT:-Kbv3G4KXHTbwqJkC8X7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b177a8f780ed7791e3de77c0815dfe5a42e5db28/pkg/sql/distsqlplan/fake_span_resolver.go#L37)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

i'd suggest moving this comment to the definition of the `fakeSpanResolver` struct, and also expanding it a bit to detail what it does - what `avgRangesPerNode` means.
And also state that each `Seek` randomizes everything.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlplan/fake_span_resolver.go, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13186#-Kbuucjz7LBm2gaBEVmm:-Kbv5bPvGL46UFac2g2k:bmxx5cw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b177a8f780ed7791e3de77c0815dfe5a42e5db28/pkg/sql/distsqlplan/fake_span_resolver.go#L80)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

why do you need a transaction here? you can do non-transactional scans...
</blockquote></details>

Duh, fixed.

---

*[pkg/sql/distsqlplan/fake_span_resolver.go, line 112 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13186#-KbuvcnqHfzPC68hAgXs:-Kbv4EVsZJb3gqB18q5E:b-bizqh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b177a8f780ed7791e3de77c0815dfe5a42e5db28/pkg/sql/distsqlplan/fake_span_resolver.go#L112)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

why `2 * ...` ?
</blockquote></details>

Added comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13186#-:-Kbv69_ODKcnMqhfMcHb:b-814g8k)*
<!-- Sent from Reviewable.io -->
